drivers/adc/adc_mcux_lpadc.c

```diff
@@ -57,7 +59,108 @@ struct mcux_lpadc_data {
 	lpadc_conv_command_config_t cmd_config[CONFIG_LPADC_CHANNEL_COUNT];
 };
 
+static int mcux_lpadc_acquisition_time_setup(const struct device *dev, uint16_t acq_time,
+					     lpadc_conv_command_config_t *cmd)
+{
+	const struct mcux_lpadc_config *config = dev->config;
+	uint32_t adc_freq_hz = 0;
+	float adc_cycle_ns = 0;
+	/* The unit of elements in the sample time array is ADCK. */
+	uint32_t acquisition_time_value = ADC_ACQ_TIME_VALUE(acq_time);
+	uint8_t acquisition_time_unit = ADC_ACQ_TIME_UNIT(acq_time);
+	uint32_t sample_time_array[] = {3, 5, 7, 11, 19, 35, 67, 131};
+	uint8_t sample_time_array_size = sizeof(sample_time_array) / sizeof(*sample_time_array);
+	uint8_t array_left_index = 0;
+	uint8_t array_right_index = sample_time_array_size - 1;
+	uint8_t array_mid_index = (array_left_index + array_right_index) / 2;
+	int left_gap = 0;
+	int right_gap = 0;
+
+	if (acquisition_time_value == ADC_ACQ_TIME_DEFAULT) {
+		return 0;
+	}
 
+	/* If the acquisition time is expressed in ADC ticks, direct match the option. */
+	if (ADC_ACQ_TIME_TICKS == acquisition_time_unit) {
+		switch (acquisition_time_value) {
+		case 3:
+			cmd->sampleTimeMode = kLPADC_SampleTimeADCK3;
+			break;
+		case 5:
+			cmd->sampleTimeMode = kLPADC_SampleTimeADCK5;
+			break;
+		case 7:
+			cmd->sampleTimeMode = kLPADC_SampleTimeADCK7;
+			break;
+		case 11:
+			cmd->sampleTimeMode = kLPADC_SampleTimeADCK11;
+			break;
+		case 19:
+			cmd->sampleTimeMode = kLPADC_SampleTimeADCK19;
+			break;
+		case 35:
+			cmd->sampleTimeMode = kLPADC_SampleTimeADCK35;
+			break;
+		case 67:
+			cmd->sampleTimeMode = kLPADC_SampleTimeADCK67;
+			break;
+		case 131:
+			cmd->sampleTimeMode = kLPADC_SampleTimeADCK131;
+			break;
+		default:
+			cmd->sampleTimeMode = kLPADC_SampleTimeADCK7;
+			return -EINVAL;
+		}
+	}
+	/* If the acquisition time is expressed in nanoseconds or microseconds, each option in the
+	 * sample time array should be converted to nanoseconds, and then find the closest
+	 * configuration option.
+	 */
+	else {
+		/* Get ADC functional clock frequency to configure ADC acquisition time. */
+		if (clock_control_get_rate(config->clock_dev, config->clock_subsys, &adc_freq_hz)) {
+			LOG_ERR("Get device LPADC operating frequency failed");
+			return -EINVAL;
+		}
+
+		adc_cycle_ns = ((float)1000000000.0 / adc_freq_hz);
+
+		for (uint8_t index = 0; index < sample_time_array_size; ++index) {
+			sample_time_array[index] =
+				(uint32_t)(sample_time_array[index] * adc_cycle_ns);
+		}
+
+		if (ADC_ACQ_TIME_MICROSECONDS == acquisition_time_unit) {
+			acquisition_time_value *= 1000;
+		}
+
+		if (sample_time_array[0] > acquisition_time_value) {
+			cmd->sampleTimeMode = kLPADC_SampleTimeADCK3;
+		} else if (sample_time_array[sample_time_array_size - 1] < acquisition_time_value) {
+			cmd->sampleTimeMode = kLPADC_SampleTimeADCK131;
+		} else {
+			while ((array_right_index - array_left_index) > 1) {
```

Comment:
I think a comment describing this search algorithm might be useful. From what I can tell we are comparing the midpoint of all possible ADC acquisition times, and based on that we search the lower or upper half of the array, repeating this process until we find the best acquisition time.

It might also be worth mentioning in a comment that the indices in the array correspond to the register values that need to be written to select specific acquisition times